### PR TITLE
EZP-29201: Additional indexes/foreign keys on the ezcontentbrowsebookmark table

### PR DESCRIPTION
**Pull request by adamwojs**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29201](https://jira.ez.no/browse/EZP-29201)
| **Bug/Improvement**| yes
| **New feature**    | yes
| **Target version** | `master` 
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

This PR introduces two additional indexes for `ezcontentbrowsebookmark`:

* Added index for the `user_id` and `node_id` columns to make `\eZ\Publish\SPI\Persistence\Bookmark\Handler::loadByUserIdAndLocationId` lookups faster
* Added index for `user_id` column to make `\eZ\Publish\SPI\Persistence\Bookmark\Handler::loadUserBookmarks`  lookups faster

and foreign keys for the `node_id`/`user_id` columns which will delete corresponding bookmarks when location/user is removed (solves EZP-29201).

_Question to reviewers: Is there any specific reason against using FKs? I'm asking because I can't find any other example in our schema_

**TODO**:
- [X] Implement feature / fix a bug.
- [X] Implement tests
- [X] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ensure if SPI cache for bookmarks is properly invalidated when subtree is removed
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by emodric** [comment]
Is there a specific reason why introduce foreign keys? I'm not against them per se, but seeing how they're not used *at all* in the schema, it seems a bit weird to add them now.

**Comment by andrerom** [comment]
@alongosz  wdyt?

**Comment by adamwojs** [comment]
> Is there a specific reason why introduce foreign keys? 

@emodric IMHO the FK is the most efficient way to meet requirement mentioned in PR description:  
when location or user is removed, related bookmarks should be removed as well

Alternative approach involves additional methods on the SPI: `deleteBySubtree`, `deleteByUserId`.

**Comment by emodric** [comment]
Sure, but using foreign keys for that (implicitly deleting related bookmarks) might not be compatible with future storage engines? @andrerom is this something that is relevant?

**Comment by ViniTou** [comment]
tbh, I am kinda against leaving deleting to cascade on database. Think we should be consistent and stick to signal slots.

no problem with FKs per se.

@mikadamczyk

**Comment by emodric** [comment]
No problem with foreign keys as well, they're awesome for enforcing the correct data, but I don't like leaving domain logic to cascade deletes, since it moves functionality outside of the application it's not clear from reading the code what is happening.

**Comment by andrerom** [comment]
> @andrerom is this something that is relevant?

Nope, we plan to use it much more in future storage engine, so this aligns ok. And for LegacyStorage engine we just need to make sure it does not break legacy, which seems to not be an issue here.

File: eZ/Publish/API/Repository/Tests/SetupFactory/Legacy.php [hunk 1]
```diff
@@ -177,11 +179,17 @@ public function insertData()
...
-        foreach ($data as $table => $rows) {
-            // Cleanup before inserting (using TRUNCATE for speed, however not possible to rollback)
-            $q = $dbPlatform->getTruncateTableSql($handler->quoteIdentifier($table));
-            $connection->executeUpdate($q);
+        foreach (array_reverse(array_keys($data)) as $table) {
+            try {
+                // Cleanup before inserting (using TRUNCATE for speed, however not possible to rollback)
+                $connection->executeUpdate($dbPlatform->getTruncateTableSql($handler->quoteIdentifier($table)));
+            } catch (DBALException | PDOException $e) {
+                // Fallback to DELETE if TRUNCATE failed (because of FKs for instance)
```
Comment (glye) [inline eZ/Publish/API/Repository/Tests/SetupFactory/Legacy.php]: Assuming here that this exception is already handled in code, and that's why it's only added in the test, not in code.

**Comment by natanael89** [comment]
IMO a test should be added so it's clear from the code that we expect the database to handle deleting related bookmarks. Perhaps a comment explaining that as well?

**Comment by adamwojs** [comment]
PR squashed and ready for review.

**Pull request merged**